File: task_cli.py
import os
import json
import shlex

file_name = "tasks.json"
id = 1

def initialize_id():
    """Initialize `id` to be the maximum ID in the JSON file + 1."""
    global id
    if os.path.exists(file_name):
        with open(file_name, 'r') as file:
            try:
                data = json.load(file)
            except json.JSONDecodeError:
                # Handle empty or invalid JSON
                data = {"tasks": []}
        if data["tasks"]:
            max_id = max(task["ID"] for task in data["tasks"])
            id = max_id + 1
            print(f"Initialized ID to {id} based on existing tasks.")
            return id
        else:
            print("No tasks found in JSON file. Using default ID = 1.")
            return id
    else:
        print("JSON file not found. Starting fresh with ID = 1.")
        return id

def add(task, mark):
    global id
    if os.path.exists(file_name):
        with open(file_name, 'r') as file:
            data = json.load(file)
        
        # Add a new task
        data["tasks"].append({"description": task, "ID": id, "mark": mark})
        id += 1
        with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
        print("New task added:",  {"ID": id - 1})
    else:
        data = {
            "tasks": [{"description": task, "ID": id, "mark": mark}]
        }
        id += 1
        with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
        print("New JSON file created with the first task:", {"description": task, "ID": id - 1, "mark": mark})


def update(task_id, new_description):
    if os.path.exists(file_name):
        with open(file_name, 'r') as file:
            data = json.load(file)
        
        # Search for the task by ID and update it
        for task in data["tasks"]:
            if task["ID"] == task_id:  # Task ID matches
                task["description"] = new_description
                print(f"Task ID {task_id} updated successfully.")
                break
        else:
            print(f"Task ID {task_id} not found.")

        # Save the updated data back to the file
        with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    else:
        print("Error: Task file not found. Please add a task first.")

def delete_task(task_id):
    if os.path.exists(file_name):
        # Load the JSON file
        with open(file_name, 'r') as file:
            data = json.load(file)
        
        # Filter out the task with the specified ID
        original_length = len(data["tasks"])
        data["tasks"] = [task for task in data["tasks"] if task["ID"] != task_id]
        
        if len(data["tasks"]) < original_length:
            # Save the updated JSON data
            with open(file_name, 'w') as file:
                json.dump(data, file, indent=4)
            print(f"Task ID {task_id} deleted successfully.")
        else:
            print(f"Task ID {task_id} not found.")
    else:
        print("Error: Task file not found. Please create a task first.")
    
def mark_in(task_id,mark):
    if os.path.exists(file_name):
        with open(file_name, 'r') as file:
            data = json.load(file)
        
        # Search for the task by ID and update it
        for task in data["tasks"]:
            if task["ID"] == task_id:  # Task ID matches
                task["mark"] = mark
                print(f"Task ID {task_id} marked in-progress successfully.")
                break
        else:
            print(f"Task ID {task_id} not found.")

        # Save the updated data back to the file
        with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    else:
        print("Error: Task file not found. Please add a task first.")

def show_list(mark):
    if os.path.exists(file_name):
        with open(file_name, 'r') as file:
            data = json.load(file)
        num = 1
        if mark == "all":
            for task in data["tasks"]:
                print(f"{num}. {task}")
                num+=1
        else:
            for task in data["tasks"]:
                if task["mark"] == mark:
                    print(f"{num}. {task}")
                    num+=1
            
    else:
        print("Error: Task file not found. Please add a task first.")
    

def tasks():
    initialize_id()
    while True:
        user_input = input("task-cli> ")  # Takes input as a string
        parsed_input = shlex.split(user_input)
        print(parsed_input)
        
        if parsed_input[0] == "add":
            mark = "todo"
            name = " ".join(parsed_input[1:])  # Handle multi-word task names
            add(name, mark)
        
        elif parsed_input[0] == "update":
            task_id = int(parsed_input[1])
            update_description = " ".join(parsed_input[2:])  # Handle multi-word descriptions
            update(task_id, update_description)
        
        elif parsed_input[0] =="delete":
            task_id = int(parsed_input[1])
            delete_task(task_id)
        
        elif parsed_input[0] =="mark-in-progress":
            task_id = int(parsed_input[1])
            mark_in(task_id,"in-progress")
            
        elif parsed_input[0] =="mark-done":
            task_id = int(parsed_input[1])
            mark_in(task_id,"done")
        
        elif parsed_input[0] =="list":
            if len(parsed_input) == 1:
                show_list("all")
            else:
                show_list(parsed_input[1])
        
        elif parsed_input[0] == "exit":
            print("Exiting task CLI.")
            break
        
        else:
            print("Unknown command. Available commands: add, update, exit.")
            
if __name__ == "__main__":
    tasks()

File: cpp/task_cli.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <nlohmann/json.hpp>
#include <sstream>
#include <limits>
#include <iomanip> // <-- Add this include for `setw`
#include "task_cli.hpp"

using json = nlohmann::json;
using namespace std;
int id = 1;

void initialize_id() {
    ifstream file(file_name);
    if (!file.is_open()) {
        cout << "JSON file not found. Starting fresh with ID = 1.\n";
        id = 1;
        return;
    }
    json data;
    file >> data;
    if (!data["tasks"].empty()) {
        int max_id = 0;
        for (auto& task : data["tasks"]) {
            max_id = max(max_id, task["ID"].get<int>());
        }
        id = max_id + 1;
    } else {
        id = 1;
    }
    cout << "Initialized ID to " << id << " based on existing tasks.\n";
    file.close();
}

void add_task(const string& description, const string& mark) {
    json data = load_data();
    data["tasks"].push_back({{"ID", id}, {"description", description}, {"mark", mark}});
    save_data(data);
    cout << "Task added: ID=" << id << ", Description=\"" << description << "\", Mark=\"" << mark << "\"\n";
    ++id;
}

void update_task(int task_id, const string& new_description) {
    json data = load_data();
    bool updated = false;
    for (auto& task : data["tasks"]) {
        if (task["ID"] == task_id) {
            task["description"] = new_description;
            updated = true;
            break;
        }
    }
    if (updated) {
        save_data(data);
        cout << "Task ID " << task_id << " updated successfully.\n";
    } else {
        cout << "Task ID " << task_id << " not found.\n";
    }
}

void delete_task(int task_id) {
    json data = load_data();
    size_t original_size = data["tasks"].size();
    data["tasks"].erase(remove_if(data["tasks"].begin(), data["tasks"].end(),
                                  [task_id](const json& task) { return task["ID"] == task_id; }),
                        data["tasks"].end());
    if (data["tasks"].size() < original_size) {
        save_data(data);
        cout << "Task ID " << task_id << " deleted successfully.\n";
    } else {
        cout << "Task ID " << task_id << " not found.\n";
    }
}

void mark_task(int task_id, const string& mark) {
    json data = load_data();
    bool updated = false;
    for (auto& task : data["tasks"]) {
        if (task["ID"] == task_id) {
            task["mark"] = mark;
            updated = true;
            break;
        }
    }
    if (updated) {
        save_data(data);
        cout << "Task ID " << task_id << " marked as \"" << mark << "\" successfully.\n";
    } else {
        cout << "Task ID " << task_id << " not found.\n";
    }
}

void show_list(const string& filter) {
    json data = load_data();
    int num = 1;
    for (const auto& task : data["tasks"]) {
        if (filter == "all" || task["mark"] == filter) {
            cout << num++ << ". ID=" << task["ID"] << ", Description=\"" << task["description"]
                 << "\", Mark=\"" << task["mark"] << "\"\n";
        }
    }
    if (num == 1) cout << "No tasks found.\n";
}

json load_data() {
    ifstream file(file_name);
    json data = {{"tasks", json::array()}};
    if (file.is_open()) {
        try {
            file >> data;
        } catch (...) {
            data["tasks"] = json::array();
        }
        file.close();
    }
    return data;
}

void save_data(const json& data) {
    ofstream file(file_name);
    file << setw(4) << data;
    file.close();
}
// #ifndef DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
// int main() {
//     initialize_id();
//     string command;
//     cout << "Task CLI initialized. Available commands: add, update, delete, mark, list, exit.\n";

//     while (true) {
//         cout << "task-cli> ";
//         getline(cin, command);
//         istringstream iss(command);
//         vector<string> tokens((istream_iterator<string>(iss)), istream_iterator<string>());

//         if (tokens.empty()) continue;

//         if (tokens[0] == "add" && tokens.size() > 1) {
//             string description = command.substr(4);
//             add_task(description);
//         } else if (tokens[0] == "update" && tokens.size() > 2) {
//             int task_id = stoi(tokens[1]);
//             string new_description = command.substr(command.find(tokens[2]));
//             update_task(task_id, new_description);
//         } else if (tokens[0] == "delete" && tokens.size() > 1) {
//             delete_task(stoi(tokens[1]));
//         } else if (tokens[0] == "mark-in-progress" && tokens.size() > 1) {
//             mark_task(stoi(tokens[1]), "in-progress");
//         } else if (tokens[0] == "mark-done" && tokens.size() > 1) {
//             mark_task(stoi(tokens[1]), "done");
//         } else if (tokens[0] == "list") {
//             if (tokens.size() > 1)
//                 show_list(tokens[1]);
//             else
//                 show_list();
//         } else if (tokens[0] == "exit") {
//             cout << "Exiting task CLI.\n";
//             break;
//         } else {
//             cout << "Unknown command. Available commands: add, update, delete, mark, list, exit.\n";
//         }
//     }

//     return 0;
// }
// #endif // DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN

